### PR TITLE
New asset format

### PR DESCRIPTION
**Pull request by McCallisterRomer**
Initial work for new asset formats. This repo now produces 2 primary targets:

nc-convert.exe
- New version of nc-tools.exe. CLI is the same, but output is in new format.
- concave collider, hull collider, and mesh are built in the new format
- cubemaps use the legacy format

NcAsset
- Library with asset type definitions and import logic for .nca files. NcEngine will replace all of its asset reading code with simple calls to this.

### CODE REVIEW

File: docs/AssetFormats.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+### Mesh Blob Format
+> Magic Number: 'MESH'
+
+| Name         | Type      | Size |
+|--------------|-----------|------|
+| extents      | Vector3   | 12   |
+| max extent   | float     | 4    |
+| vertex count | u64       | 8    |
+| index count  | u64       | 8    |
+| vertext list | Vector3[] | -    |
```
Comment (McCallisterRomer) [inline docs/AssetFormats.md]: Oops, I need to update this to `MeshVertex`.

File: include/ncasset/AssetTypes.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+#include <array>
+#include <vector>
+
+namespace nc::asset
+{
+enum class AssetType
+{
+    AudioClip,
+    Cubemap,
```
Comment (jaremieromer) [inline include/ncasset/AssetTypes.h]: Will you change to CubeMap? We've settled on treating it as two words - the jury seems to be split on the net.

File: include/ncasset/AssetTypes.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+struct Texture
+{
+    uint32_t width;
+    uint32_t height;
+    std::vector<unsigned char> pixels;
+};
+
+struct Cubemap
```
Comment (jaremieromer) [inline include/ncasset/AssetTypes.h]: Will you change to CubeMap? We've settled on treating it as two words - the jury seems to be split on the net.

File: docs/AssetFormats.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+## Nc Asset
+-----------
+An asset can exist on its own in a .nca file, or as an entry in a .ncp package. The data layout is identical in either case. Each nca consists of a header that is common between all asset types and
+an asset blob, the layout of which is determined by the asset type.
+
+### .nca File Format
+| Name         | Type    | Size | Note |
+|--------------|---------|------|------
+| magic number | string  | 4    | non-null-terminated string identifying the asset type
```
Comment (jaremieromer) [inline docs/AssetFormats.md:27]: can you change to identifier, uid, or something else? 'magic number' is actually a negative thing - it's a term used to represent an anti-pattern

File: source/common/Deserialize.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
+#include "Deserialize.h"
```
Comment (jaremieromer) [inline source/common/Deserialize.cpp:1]: Nice and neat and concise! I love it

File: source/common/NcaHeader.h [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <cstddef>
+
+namespace nc::asset
+{
+/** @brief Identifiers for asset blobs in .nca files. */
+struct MagicNumber
```
Comment (jaremieromer) [inline source/common/NcaHeader.h:8]: See comment about magic number

Comment (jaremieromer) [reply source/common/NcaHeader.h:8]: I'd even prefer ATI -asset type identifier

File: docs/AssetFormats.md (hunk 4 above)
Comment (jaremieromer) [reply docs/AssetFormats.md:27]: Per Wikipedia: 
The term magic number or magic constant refers to the [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern) of using numbers directly in source code.

File: source/common/Serialize.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#include "Serialize.h"
+#include "NcaHeader.h"
+
+#include <ostream>
+
+namespace
+{
+void WriteHeader(std::ostream& stream, const char* magicNumber, size_t assetId, size_t size)
```
Comment (jaremieromer) [inline source/common/Serialize.cpp:8]: This file is also great!

File: source/ncconvert/analysis/GeometryAnalysis.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+#include <type_traits>
+
+namespace
+{
+auto CompareSquareMagnitudes(const nc::Vector3& lhs, const nc::Vector3& rhs)
+{
+    return nc::SquareMagnitude(lhs) < nc::SquareMagnitude(rhs);
+}
+
+auto PositionProjection(const nc::asset::MeshVertex& vertex) -> const nc::Vector3&
```
Comment (jaremieromer) [inline source/ncconvert/analysis/GeometryAnalysis.cpp:19]: Why is this called PositionProjection?

File: source/ncconvert/analysis/GeometryAnalysis.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace nc
+{
+struct Triangle;
+struct Vector3;
+
+namespace asset
+{
+struct MeshVertex;
+}
```
Comment (jaremieromer) [inline source/ncconvert/analysis/GeometryAnalysis.h]: Should have // namespace asset here for consistency

File: test/common/Serialize_unit_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    ASSERT_EQ(expectedAsset.indices.size(), actualAsset.indices.size());
+
+    for(auto i = 0u; i < expectedAsset.vertices.size(); ++i)
+    {
+        const auto& e = expectedAsset.vertices[i];
+        const auto& a = actualAsset.vertices[i];
+        EXPECT_EQ(e, a);
+    }
+
+    // EXPECT_TRUE(std::equal(expectedAsset.vertices.cbegin(),
```
Comment (jaremieromer) [inline test/common/Serialize_unit_tests.cpp]: Does this commented out code need to hang around?

**Review by jaremieromer — APPROVED** [review]
NICE!

File: include/ncasset/AssetTypes.h (hunk 2 above)
Comment (McCallisterRomer) [reply include/ncasset/AssetTypes.h]: Fixed

File: source/ncconvert/analysis/GeometryAnalysis.cpp (hunk 8 above)
Comment (McCallisterRomer) [reply source/ncconvert/analysis/GeometryAnalysis.cpp:19]: The standard algorithms in the `ranges` namespace have an optional projection function that transforms input data before passing to the actual operation, and this guy is used for those algorithms. All of the mesh analysis stuff only cares about positional data, so we 'project' the `MeshVertex` to just its position. Added a comment here.

File: test/common/Serialize_unit_tests.cpp (hunk 10 above)
Comment (McCallisterRomer) [reply test/common/Serialize_unit_tests.cpp]: Nope!

**Pull request merged**